### PR TITLE
feat: toBeOnTheScreen() matcher

### PR DESCRIPTION
**Pull request by mdjastrzebski**
**What**:

Jest matcher `toBeOnTheScreen()`:
* it is re-implementation of Jest DOM matcher named `toBeInTheDocument()`
* it checks whether given element is still part of the element tree

**Why**:

Before this query RNTL suggested using `toBeTruthy()` predicate on query results. It generally works fine for query that is issued just before making assertion, but it is not able to verify that a captured element has been removed from element tree.

**How**:

Adapted code from Jest DOM matchers.

**Checklist**:

- [x] Documentation added to the
      [docs](https://github.com/testing-library/jest-native/README.md)
- [x] Typescript definitions updated
- [x] Tests
- [x] Ready to be merged 

@pierrezimmermannbam, @AugustinLF, @MattAgn wdyt?

### CODE REVIEW

**Comment by mdjastrzebski** [comment]
Some points for discussion:
1. Whether we should use RTL-compat name when there is no document concept in RN. Maybe we could pick another meaningful name like `toBeInTheElementTree()`, `toBeInTheComponentTree()`, `toBeMounted()`, `toBeRendered()`
2. If we pick another name, whether we should provide RTL-compat alias `toBeInTheDocument()`

I also think that it would make sense to avoid explicit dep on RNTL, and rather do `require` import and show warning if the import is not found. Indicating that user should have RNTL in minimal version xx to use this matcher. This info would be also put into docs.

**Comment by pierrezimmermannbam** [comment]
I'm not sure I really understand the use case if this matcher. If I understand correctly, whenever we update what is rendered, toBeInDocument() will throw since we compare the root elements and don't check wether the element is still rendered or not. 

So for instance this would fail if I understand correctly : 

```tsx
render(<Text>hello</Text>);

const element = screen.getByText('hello');
screen.update(<Text>hello</Text>)

expect(element).toBeInTheDocument();
```

If that's really the case I can't think of a situation where I'd want to use this matcher, but maybe I'm missing something ? What are the use cases you envision for this matcher ?

Regarding the matcher naming I think it should be changed to be more meaningful. I'm not sure I like any of the names you suggested as they are too react specific imo, what about toBeInTheScreen ?

**Comment by mdjastrzebski** [comment]
@pierrezimmermannbam For query issue at the same time as matcher this is behaving the same as currently recommend `toBeTruthy()` matcher. However, this matcher allows you to test whether element captured as variable before some component rerender is still attached to the element tree, as in such case the capture variable would be still truthy but could be not attached to the element tree any more.

Re test: that test would actually pass, as `ReactTestInstance` is actually actually a wrapper on an underlying React fiber, which remains the same.

This query is adaptation of `toBeInTheDocument` from RTL, which is their recommended query for element existence check:
```ts
  expect(getByText('Hello, world!')).toBeInTheDocument()
```

File: src/to-be-in-the-document.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  const pass = element === null ? false : screen.container === getRootElement(element);
+
+  const errorFound = () => {
+    return `expected document not to contain element but found:\n${printElement(element)}`;
+  };
+
+  const errorNotFound = () => {
+    return `element could not be found in the document`;
```
Comment (thymikee) [inline src/to-be-in-the-document.ts:18]: A good error message has a recovery hint. Something like "Please make sure you rendered the element." would be welcome

File: package.json [hunk 2]
```diff
@@ -64,8 +64,7 @@
   "peerDependencies": {
     "react": ">=16.0.0",
     "react-native": ">=0.59",
-    "react-test-renderer": ">=16.0.0",
-    "@testing-library/react-native": ">=11.0.0"
```
[the remark this answers is not in the document]
Comment (mdjastrzebski) [reply package.json]: Not sure what do you mean? I've added RNTL as peer deps in the initial version of this PR, and now I am using checked `require` just in this query.

**Comment by mdjastrzebski** [comment]
@thymikee @pierrezimmermannbam I've tweaked the code a bit. You both mentioned naming so wanted to brainstorm the naming options. Pls vote with thumb ups:

**Comment by mdjastrzebski** [comment]
1. `toBeOnTheScreen()`

**Comment by mdjastrzebski** [comment]
3. `toBeMounted()`

File: extend-expect.d.ts [hunk 3]
```diff
@@ -12,11 +12,12 @@ declare global {
       toHaveTextContent(text: string | RegExp, options?: { normalizeWhitespace: boolean }): R;
       toBeEnabled(): R;
       toHaveStyle(style: StyleProp<ViewStyle | TextStyle | ImageStyle>): R;
+      toBeVisible(): R;
```
Comment (pierrezimmermannbam) [inline extend-expect.d.ts:15]: it seems this line has been duplicated

**Comment by pierrezimmermannbam** [comment]
> Re test: that test would actually pass, as `ReactTestInstance` is actually actually a wrapper ("forwarding delegate") to an underlying React fiber, which remains the same :-)

This part is still a bit confusing to me, I see better how this matcher behaves but I don't understand fully yet why it does, could you point me to some documentation where this is further detailed ?

**Comment by mdjastrzebski** [comment]
@pierrezimmermannbam Good point, this subject is not well documented. 

You can check `ReactTestInstance` [source](https://github.com/facebook/react/blob/c54e3541b2fb6251bd7cd399e77037220b47de2e/packages/react-test-renderer/src/ReactTestRenderer.js#LL283C7-L283C24). Inside you can see that RTI is not an immutable object, but rather holds a reference to a mutable React fiber object. All RTI props are actually computed and work on current fiber tree. 

There is also a `WeakMap` that is responsible for [keeping cached fiber to RTI mapping](https://github.com/facebook/react/blob/c54e3541b2fb6251bd7cd399e77037220b47de2e/packages/react-test-renderer/src/ReactTestRenderer.js#LL614C3-L614C3).

That's basically it.

**Comment by pierrezimmermannbam** [comment]
@mdjastrzebski thank you very much I'll look into it !

**Review by pierrezimmermannbam — APPROVED** [review]

**Review by thymikee — APPROVED** [review]

**Review by AugustinLF — APPROVED** [review]

**Pull request merged**